OpenRA.Mods.Common/Traits/World/ShroudRenderer.cs

```diff
@@ -197,108 +215,162 @@ Edges GetObserverEdges(CPos p)
-				var fogged = GetEdges(uv, visibleUnderFog);
-				var shroudTile = tiles[uv];
-				var variant = shroudTile.Variant;
-				shroudTile.Shroud = GetTile(shroudSprites, shrouded, variant);
-				shroudTile.Fog = GetTile(fogSprites, fogged, variant);
-				tiles[uv] = shroudTile;
-			}
+			// Mark this cell and its 8 neighbors as being out of date.
+			// We don't want to do anything more than this for several performance reasons:
+			// - If the cells remain off-screen for a long time, they may change several times before we next view
+			// them, so calculating their new vertices is wasted effort since we may recalculate them again before we
+			// even get a chance to render them.
+			// - Cells tend to be invalidated in groups (imagine as a unit moves, it advances a wave of sight and
+			// leaves a trail of fog filling in behind). If we recalculated a cell and its neighbors when the first
+			// cell in a group changed, many cells would be recalculated again when the second cell, right next to the
+			// first, is updated. In fact we might do on the order of 3x the work we needed to!
+			shroudDirty[cell + new CVec(-1, -1)] = true;
+			shroudDirty[cell + new CVec(0, -1)] = true;
+			shroudDirty[cell + new CVec(1, -1)] = true;
+			shroudDirty[cell + new CVec(-1, 0)] = true;
+			shroudDirty[cell] = true;
+			shroudDirty[cell + new CVec(1, 0)] = true;
+			shroudDirty[cell + new CVec(-1, 1)] = true;
+			shroudDirty[cell + new CVec(0, 1)] = true;
+			shroudDirty[cell + new CVec(1, 1)] = true;
 		}
 
 		void UpdateNullShroud()
 		{
-			foreach (var cell in map.Cells)
+			// Cache the whole shroud when there are no observers, since it will never change.
+			foreach (var uv in map.Cells.MapCoords)
 			{
-				var edges = GetObserverEdges(cell);
-				var shroudTile = tiles[cell];
-				var variant = shroudTile.Variant;
-				shroudTile.Shroud = GetTile(shroudSprites, edges, variant);
-				shroudTile.Fog = GetTile(fogSprites, edges, variant);
-				tiles[cell] = shroudTile;
+				var offset = VertexArrayOffset(uv);
+				var edges = GetObserverEdges(uv.ToCPos(map));
+				var tileInfo = tileInfos[uv];
+				CacheTile(uv, offset, edges, tileInfo, shroudSprites, shroudVertices, shroudPalette, shroudSpriteLayer);
+				CacheTile(uv, offset, edges, tileInfo, fogSprites, fogVertices, fogPalette, fogSpriteLayer);
 			}
 		}
 
-		public void RenderShroud(WorldRenderer wr, Shroud shroud)
+		void Render(CellRegion visibleRegion)
 		{
-			Update(shroud, wr.Viewport.VisibleCells);
+			var renderRegion = CellRegion.Expand(visibleRegion, 1).MapCoords;
+
+			if (currentShroud == null)
+			{
+				RenderObserverShroud(renderRegion);
+				return;
+			}
 
-			foreach (var uv in CellRegion.Expand(wr.Viewport.VisibleCells, 1).MapCoords)
+			RenderPlayerShroud(visibleRegion, renderRegion);
+		}
+
+		void RenderObserverShroud(CellRegion.MapCoordsRegion renderRegion)
+		{
+			// The observer shroud is always fully cached, so we can just render straight from the cache.
```

Comment:
My intention was to convey the fact that the observer shroud is fixed. We generate all the tiles (i.e. cache) it right away (see UpdateNullShroud above) and so when we need to render it, we can do so without having to check for dirty tiles.

Would something more akin to "The observer shroud never changes and we have cached all the tiles ahead of time - therefore we can render all the cached tiles without checking for any dirty tiles" be clearer?